hit-testing-explainer.md

```diff
@@ -0,0 +1,275 @@
+
+The `XRInputSourcesChangeEvent.getHitTestSource()` function will only return a `XRHitTestSource` object when passed an `XRInputSource` from the `XRInputSourcesChangeEvent.added` attribute. Any other parameter will cause a `DOMException` to be thrown.
+
+```js
+let hitTestSources = {};
+
+function onInputSourcesChange(event) {
+  xrInputSources = event.session.getInputSources();
+
+  foreach (inputSource of event.removed) {
+    delete hitTestSources[inputSource];
+  }
+
+  foreach (inputSource of event.added)
+    hitTestSources[inputSource] = event.getHitTestSource(inputSource);
+  }
+
+  updatePreferredInputSource();
+}
+```
+
+### Hit test results
+To get synchronous hit test results for a particular frame, developers call `XRFrame.getHitTestResults()` passing in a `XRHitTestSource` as the `hitTestSource` parameter. This function will return a `FrozenArray<XRHitTestResult>` in which `XRHitTestResult`s are ordered by distance from the `XRHitTestSource`, with the nearest in the 0th position. If no results exist, the array will have a length of zero. Each entry in the array will have a `hitTestOptions` attribute filled in with `XRHitTestOptions` of the `XRHitTestSource` used to find the result. Each entry will also have a `transform` attribute that represents the result's location in 3D space. If no value is provided for `relativeTo`, transforms will be defined in the coordinate system of the `hitTestOptions.space`. Otherwise, transforms will be defined in the coordinate system of the `relativeTo`. If `relativeTo` is present and cannot be located relative to `hitTestOptions.space` on the current frame, the function will return `null`.
+
+```js
+function updateScene(timestamp, xrFrame) {
+  // Scene update logic ...
+
+  let hitTestResults = xrFrame.getHitTestResults(hitTestSources[preferredInputSource], xrReferenceSpace);
+  if (hitTestResults && hitTestResults.length > 0) {
+    // Do something with the results
+  }
+
+  // Other scene update logic ...
+}
+```
+
+On occasion, developers may want hit test results for the current frame even if they have not already created an `XRHitTestSource` to subscribe to the results. For example, when a virtual object needs to bounce off a real-world surface, a single hit-test result can be requested. The results will be delivered asynchronously, though they will be accurate for the frame on which the request was made. Otherwise, `requestAsyncHitTestResults()` shares the behavior of `getHitTestResults()` as described above.
```

Comment:
Is the behavior of "request now, receive results later, but results are based on data from previous frame" what apps will actually want?  Do we have a concrete scenario for wanting that?

I see two coherent paths for the following scenario:
> At time A, the user presses a button.  After some amount of wireless latency, the UA receives the press and sets controller state, available to the app for an `XRFrame` with forward-predicted time B.  At the frame for forward-predicted time C, the app gets results that match either...
* Time A, when the press actually occurred.  This would be useful for hit-testing what a controller was pointing at back when the press occurred, (available through `XRInputSourcesChangeEvent.getHitTestSource()`)  _or_
* Time C, for the forward-predicted time at which the app will target its current frame.  This would be useful for a view-relative reticle. (available through `XRFrame.getHitTestResults(hitTestSource)`)

Time B seems arbitrary relative to the user's actions, and so going out of our way to return data for Time B but not Time A seems of low value.  This makes me question if there is a strong value now for the arbitrary hit-test. (`XRFrame.requestAsyncHitTestResults`)

Reply:
A relevant nuance is that what XR APIs generally allow time-indexing of is **dynamic** data, for things like heads and controllers that are actually moving.  In contrast, for **static** and **semi-static** data, such as planes and meshes, APIs generally only expose the latest snapshot of their poses and geometry.

Unless a plane itself is being posed smoothly every frame in a forward-predicted manner, the following may be a reasonable balance that gets apps what they need while imposing a minimal constraint on UA architecture:
* Raycast **from** the pose of the `XRSpace` at the exact historical or future time requested in the `hitTest` call.
* Raycast **against** the world geometry that is current at the moment when the data is delivered.